### PR TITLE
Python: Strip background and continuation_token options after background response completes in tool loop

### LINKED ISSUE #5394: Python: [Bug]: `background=True` causes infinite tool-call loop, tool-result submissions inherit background mode
> **Correction (updated after deeper investigation):**
> My initial diagnosis blamed `background=True` leaking into tool-result submissions. That was **wrong**. The actual culprit is **`continuation_token`** leaking into `mutable_options` across tool-loop iterations, causing the SDK to repeatedly `GET` the same completed response instead of `POST`ing tool results.
>
> The HTTP log is the smoking gun: **1 POST + ~40 GETs all to the same `response_id`**, no tool-result POSTs ever happen. See the updated Description below. A runtime monkeypatch that fixes it is verified working and included at the bottom.

`background=True` + function tools: SDK loops on `responses.retrieve()` of the same response, never POSTs tool results, returns empty text

## Description

**SDK version:** `agent-framework-core==1.1.0`, `agent-framework-openai==1.1.0`
**Python version:**:  `3.14`

When an agent runs with `background=True` and has local function tools, the SDK gets stuck retrieving the same completed background response on every tool-loop iteration. Tool results are never POSTed back. After `max_iterations` (default 40) the loop exits and `response.text` is empty.

### What actually happens

Verified from `httpx` INFO logs: the traffic is 1 POST followed by roughly 40 GETs, and every GET hits the same `response_id`.

1. Caller does `agent.run(messages, options={background: True})`. This POSTs and gets back a `continuation_token`.
2. Caller polls with `agent.run(session=session, options={continuation_token: X})` in a loop.
3. Eventually `RawOpenAIChatClient._inner_get_response()` sees `continuation_token` in options and calls `client.responses.retrieve(response_id)`. The retrieve returns a completed response with `function_call` output items.
4. `FunctionInvocationLayer` executes the tools locally and enters the next tool-loop iteration.
5. `mutable_options` (the shared dict inside `FunctionInvocationLayer.get_response()`) still contains `continuation_token`, so the next `super_get_response()` call does another `responses.retrieve()` to the same `response_id`.
6. Same completed response, same `function_call` items. Tools run again.
7. Repeat up to 40 times. No tool result is ever POSTed. The model never gets the tool outputs. Final response text is empty.

### Root cause

`FunctionInvocationLayer.get_response()` copies all options into a `mutable_options` dict (around line 59 of `_tools.py`) and reuses that same dict for every `super_get_response()` call in the tool loop. When the caller's polling code passes `continuation_token` in options, it leaks into every iteration. `_inner_get_response()` sees it and keeps doing GET retrieves instead of POSTing tool results.

`background=True` has the same problem. If only `continuation_token` were fixed, tool-result POSTs would still start new background jobs instead of running synchronously.

For comparison, `conversation_id` is handled correctly: it gets updated on `mutable_options` by `_update_continuation_state` after each response. `continuation_token` needs the inverse treatment, it should be removed from `mutable_options` once the background job returns a completed response.

## Code Sample

Minimal reproduction. One tool, background mode, structured output.

```python
import asyncio
from agent_framework.openai import OpenAIChatClient, OpenAIChatOptions
from agent_framework import AgentSession, Message, tool
from pydantic import BaseModel


@tool(name="search", description="Search documents for information")
def search(query: str) -> str:
    return f"Document contains info about: {query}"


class Output(BaseModel):
    answer: str


client = OpenAIChatClient(
    api_key="YOUR_KEY",
    model="gpt-4.1",
    azure_endpoint="https://YOUR_ENDPOINT",
    api_version="preview",
)

agent = client.as_agent(
    instructions="Use the search tool to find info, then answer.",
    name="repro_agent",
    tools=[search],
    default_options=OpenAIChatOptions(response_format=Output),
)


async def reproduce():
    session = AgentSession()
    messages = [Message(role="user", contents=["Search for 'project name'"])]

    response = await agent.run(
        messages, session=session, options=OpenAIChatOptions(background=True)
    )

    poll_count = 0
    while response.continuation_token is not None:
        poll_count += 1
        await asyncio.sleep(2.0)
        response = await agent.run(
            session=session,
            options=OpenAIChatOptions(continuation_token=response.continuation_token),
        )

    print(f"Polls: {poll_count}")
    print(f"response.text: '{response.text}'")  # empty


asyncio.run(reproduce())
```

**Expected:** `response.text` contains valid JSON matching `Output`. Tool called roughly once.

**Actual:** `response.text` is empty. Tool called 40+ times with identical arguments.

## Error Messages / Stack Traces

No exception is raised. The bug is silent. The SDK exhausts all 40 iterations and returns an empty response.

### HTTP traffic (before fix)

From `httpx` INFO logs. Note the identical `response_id` on every GET.

```
POST .../openai/v1/responses?api-version=preview "HTTP/1.1 200 OK"
  Poll 1...
GET  .../openai/v1/responses/resp_058e072487625bc8...?api-version=preview "HTTP/1.1 200 OK"
  Poll 2...
GET  .../openai/v1/responses/resp_058e072487625bc8...?api-version=preview "HTTP/1.1 200 OK"
  Poll 3...
GET  .../openai/v1/responses/resp_058e072487625bc8...?api-version=preview "HTTP/1.1 200 OK"
  (~40 more GETs to the SAME resp_058e072487625bc8... id)
GET  .../openai/v1/responses/resp_058e072487625bc8...?api-version=preview "HTTP/1.1 200 OK"
```

Exactly one POST. Zero tool-result POSTs. Every other request is a GET retrieve of the same `response_id`. This is what proves the SDK is stuck retrieving the same completed response instead of looping through fresh tool-result submissions.

### SDK log when the loop exhausts

```
Maximum iterations reached (40). Requesting final response without tools.
```

### Observed output

```
--- Result (polls=3) ---
Text length: 0
Text (raw):

Tool calls made: 120
  {'tool': 'search_documents', 'query': 'prosjektnavn adresse byggherre...', 'hits': 0}
  {'tool': 'search_documents', 'query': 'tilbudsinnbydelse prosjekt byggherre adresse', 'hits': 0}
  {'tool': 'search_documents', 'query': 'konkurransegrunnlag prosjektbeskrivelse...', 'hits': 0}
  (same 3 queries repeated 40 times = 120 calls)

*** EMPTY RESPONSE TEXT ***
```

The same tool calls repeat identically on every iteration because the SDK retrieves the same completed background response every time. Tool results are never POSTed, so the model never produces a final answer.

### Affected code paths

| Location | What it does |
|---|---|
| `agent_framework._tools.FunctionInvocationLayer.get_response()` around line 59 | `mutable_options = dict(options)`. Copies `continuation_token` and `background` into the dict shared across all loop iterations. |
| `agent_framework._tools.FunctionInvocationLayer.get_response()` around line 189 | `super_get_response(options=mutable_options)`. Reuses the same dict on every iteration, including after a retrieve has already returned a completed response. |
| `agent_framework_openai._chat_client.RawOpenAIChatClient._inner_get_response()` | Reads `continuation_token` from options. If set, does `client.responses.retrieve(response_id)` instead of POSTing. Never strips it after a completed retrieve. |

### Workaround

Do not combine background mode with tools:

```python
use_background = tools is None
```

### Runtime monkeypatch (verified)

Patch `RawOpenAIChatClient._inner_get_response` so that once a `responses.retrieve()` returns a non-in-progress response, it removes `continuation_token` and `background` from the shared options dict. The `FunctionInvocationLayer` passes the same dict reference across iterations, so mutating it in place is enough to make the next iteration POST tool results properly.

```python
from agent_framework_openai._chat_client import RawOpenAIChatClient

if not hasattr(RawOpenAIChatClient, "_true_original_inner_get_response"):
    RawOpenAIChatClient._true_original_inner_get_response = RawOpenAIChatClient._inner_get_response


def _patched_inner_get_response(self, *, messages, options, stream=False, **kwargs):
    continuation_token = options.get("continuation_token")
    if not stream and continuation_token is not None:
        async def _run():
            validated_options = await self._validate_options(options)
            try:
                raw_response = await self.client.responses.retrieve(continuation_token["response_id"])
            except Exception as ex:
                self._handle_request_error(ex)
            chat_response = self._parse_response_from_openai(raw_response, options=validated_options)
            if chat_response.continuation_token is None and isinstance(options, dict):
                options.pop("continuation_token", None)
                options.pop("background", None)
            return chat_response
        return _run()
    return RawOpenAIChatClient._true_original_inner_get_response(
        self, messages=messages, options=options, stream=stream, **kwargs
    )


RawOpenAIChatClient._inner_get_response = _patched_inner_get_response
```

### HTTP traffic with the monkeypatch applied

```
POST .../openai/v1/responses?api-version=preview "HTTP/1.1 200 OK"
  Poll 1...
GET  .../openai/v1/responses/resp_01a797ec4c3e877b...?api-version=preview "HTTP/1.1 200 OK"
  Poll 2...
GET  .../openai/v1/responses/resp_01a797ec4c3e877b...?api-version=preview "HTTP/1.1 200 OK"
  Poll 3...
GET  .../openai/v1/responses/resp_01a797ec4c3e877b...?api-version=preview "HTTP/1.1 200 OK"
  [MONKEYPATCH] stripped continuation_token+background from mutable_options
POST .../openai/v1/responses?api-version=preview "HTTP/1.1 200 OK"
POST .../openai/v1/responses?api-version=preview "HTTP/1.1 200 OK"
  (roughly 16 tool-result POSTs)
POST .../openai/v1/responses?api-version=preview "HTTP/1.1 200 OK"
```

Result: valid JSON response matching the schema, tools called with varied queries (normal agent behavior) rather than the same 3 queries repeated 40 times.

### PR DESCRIPTION
**Pull request by chetantoshniwal**
### Motivation and Context

When `background=True` is set, the tool-call loop continues to pass `background` and `continuation_token` options on subsequent iterations. This causes tool-result submissions to start new background jobs or re-fetch the same completed response, resulting in an infinite loop.

Fixes #5394

### Description

The root cause is that `mutable_options` retains `background` and `continuation_token` keys across tool-loop iterations in `FunctionInvocationLayer`. The fix strips both keys from `mutable_options` once a background response completes (i.e., `response.continuation_token is None`), ensuring subsequent tool-result POST requests are sent as normal synchronous calls. This is applied in both the streaming and non-streaming code paths.

### Contribution Checklist

- [X] The code builds clean without any errors or warnings
- [X] The PR follows the [Contribution Guidelines](https://github.com/microsoft/agent-framework/blob/main/CONTRIBUTING.md)
- [X] All unit tests pass, and I have added new tests where possible
- [ ] **Is this a breaking change?** If yes, add "[BREAKING]" prefix to the title of the PR.

---

> Note: PR autogenerated by chetantoshniwal's agent

### CODE REVIEW

File: python/packages/core/tests/core/test_function_invocation_logic.py [hunk 1]
```diff
@@ -3807,3 +3807,97 @@ def empty_request(task: str) -> str:
...
+            "tool_choice": "auto",
+            "background": True,
+        },
+    )
+
+    assert chat_client_base.call_count >= 2
+    assert recorded_options[0].get("background") is True, "First call should include background=True"
+    assert not recorded_options[1].get("background", False), (
+        "background must be stripped after the background response completes"
+    )
```
Comment (chetantoshniwal) [inline python/packages/core/tests/core/test_function_invocation_logic.py:4027]: Both new tests only exercise the non-streaming `get_response` path. The identical fix was also applied in the streaming `_stream()` generator (lines 2507-2509 of `_tools.py`). Please add at least one streaming-mode test to cover that path. Additionally, consider a negative test where the `ChatResponse` has a non-None `continuation_token` to verify the options are preserved when the background job is still in progress.

File: python/packages/core/agent_framework/_tools.py [hunk 2]
```diff
@@ -2349,6 +2349,12 @@ async def _get_response() -> ChatResponse[Any]:
                         options=mutable_options,
                     )
 
+                    # Once a background response completes, strip polling/background
+                    # options so subsequent tool-loop iterations POST results normally.
+                    if response.continuation_token is None:
```
Comment (chetantoshniwal) [inline python/packages/core/agent_framework/_tools.py]: This condition fires for every response in every tool-loop iteration, not just completed background responses, because `ChatResponse.continuation_token` defaults to `None`. The `pop` calls are no-ops when the keys aren't present so correctness is preserved, but tightening the guard makes the intent explicit.

```suggestion
                if response.continuation_token is None and (
                        "continuation_token" in mutable_options or "background" in mutable_options
                ):
```

File: python/packages/core/agent_framework/_tools.py [hunk 3]
```diff
@@ -2496,6 +2502,12 @@ async def _stream() -> AsyncIterable[ChatResponseUpdate]:
                     options=mutable_options,
                 )
 
+                # Once a background response completes, strip polling/background
+                # options so subsequent tool-loop iterations POST results normally.
+                if response.continuation_token is None:
```
Comment (chetantoshniwal) [inline python/packages/core/agent_framework/_tools.py]: Same over-broad guard as the non-streaming path: fires on every response, not just completed background responses. Tighten for clarity.

```suggestion
                if response.continuation_token is None and (
                    "continuation_token" in mutable_options or "background" in mutable_options
                ):
```

**Review by chetantoshniwal — COMMENTED** [review]
## Automated Code Review

**Reviewers:** 4 | **Confidence:** 89%

### ✓ Correctness
> The change correctly strips `continuation_token` and `background` from `mutable_options` after a background response completes (i.e., when `response.continuation_token is None`), preventing subsequent tool-loop iterations from re-polling the same completed response or inadvertently starting new background jobs. The fix is applied symmetrically in both the non-streaming and streaming paths. The `dict.pop(key, None)` calls are safe no-ops when the keys don't exist (normal non-background responses). The two new tests correctly verify the behavior by monkey-patching `_get_non_streaming_response` to record options snapshots at each call.

### ✓ Security Reliability
> This PR fixes a reliability bug where `continuation_token` and `background` options leaked across tool-loop iterations after a background response completed. The fix is correct, minimal, and applied consistently in both the non-streaming and streaming paths. The `dict.pop(..., None)` calls are safe no-ops when the keys are absent (i.e., for normal non-background responses where `continuation_token` defaults to `None`). The tests properly verify both keys are stripped. No security or reliability issues found.

### ✗ Test Coverage
> The two new tests cover the non-streaming path for stripping `continuation_token` and `background` individually, which directly validates the new logic. However, there are notable gaps: (1) the identical code added in the streaming path (lines 2507-2509) has no test coverage at all, (2) there is no negative/guard test verifying that these options are *preserved* when `continuation_token` is still non-None (i.e., the background job is still in progress), and (3) there is no test exercising the case where both `continuation_token` and `background` are present simultaneously. The streaming gap is the most significant since the same bug fix was applied to two code paths but only one is tested.

### ✓ Design Approach
> The fix correctly addresses a real bug: `continuation_token` and `background` from the caller's initial `mutable_options` were leaking into subsequent tool-loop iterations, causing poll-instead-of-POST behavior. The approach is consistent with how the rest of `mutable_options` is managed in the loop (e.g., `tool_choice` is similarly forcibly overwritten). Two minor design issues are worth noting. First, the guard `if response.continuation_token is None` fires on every single response — not just completed background responses — because `continuation_token` defaults to `None` on `ChatResponse`. The pop calls are no-ops in non-background flows so there is no correctness issue, but a tighter guard (checking whether the keys are actually in `mutable_options` first) would make the intent explicit and avoid unnecessary coupling between response state and options cleanup. Second, the tests only exercise the trivial case where the very first response has `continuation_token=None`, so the background job 'completes' immediately. The more important invariant — that options survive intermediate polling iterations where `continuation_token` remains non-None, and are only stripped when it finally becomes None — is not tested.

### Flagged Issues
- [ ] The streaming code path (lines 2507-2509 in _tools.py) has identical new logic to the non-streaming path but has zero test coverage. A streaming-mode test should be added to verify the same stripping behavior.

### Suggestions
- Tighten the guard `if response.continuation_token is None` to `if response.continuation_token is None and ('continuation_token' in mutable_options or 'background' in mutable_options)` in both the non-streaming and streaming paths, making the intent explicit and avoiding unnecessary coupling between response state and options cleanup on every iteration.
- Add a test simulating multiple polling iterations — where early responses have a non-None `continuation_token` and only the final response has `continuation_token=None` — to verify that options are preserved across intermediate iterations and stripped only on completion.
- Add a test where both `continuation_token` and `background` are present in options simultaneously to verify both are stripped together when the response completes.
- Consider combining the two existing tests into a single parameterized test (e.g., ``@pytest`.mark.parametrize`) to reduce duplication, since the test bodies are nearly identical.

---
*Automated review by chetantoshniwal's agents*

**Comment by moonbox3** [comment]
<a href="https://github.com/microsoft/agent-framework/blob/c33ec18ae94b050922f73d66a130583346e57dc5/README.md">Python Test Coverage</a><details><summary>Python Test Coverage Report • </summary><table><tr><th>File</th><th>Stmts</th><th>Miss</th><th>Cover</th><th>Missing</th></tr><tbody><tr><td colspan="5"><b>packages/core/agent_framework</b></td></tr><tr><td>&nbsp; &nbsp;<a href="https://github.com/microsoft/agent-framework/blob/c33ec18ae94b050922f73d66a130583346e57dc5/packages/core/agent_framework/_tools.py">_tools.py</a></td><td>987</td><td>83</td><td>91%</td><td><a href="https://github.com/microsoft/agent-framework/blob/c33ec18ae94b050922f73d66a130583346e57dc5/packages/core/agent_framework/_tools.py#L219-L220">219&ndash;220</a>, <a href="https://github.com/microsoft/agent-framework/blob/c33ec18ae94b050922f73d66a130583346e57dc5/packages/core/agent_framework/_tools.py#L395">395</a>, <a href="https://github.com/microsoft/agent-framework/blob/c33ec18ae94b050922f73d66a130583346e57dc5/packages/core/agent_framework/_tools.py#L397">397</a>, <a href="https://github.com/microsoft/agent-framework/blob/c33ec18ae94b050922f73d66a130583346e57dc5/packages/core/agent_framework/_tools.py#L410">410</a>, <a href="https://github.com/microsoft/agent-framework/blob/c33ec18ae94b050922f73d66a130583346e57dc5/packages/core/agent_framework/_tools.py#L435-L437">435&ndash;437</a>, <a href="https://github.com/microsoft/agent-framework/blob/c33ec18ae94b050922f73d66a130583346e57dc5/packages/core/agent_framework/_tools.py#L445">445</a>, <a href="https://github.com/microsoft/agent-framework/blob/c33ec18ae94b050922f73d66a130583346e57dc5/packages/core/agent_framework/_tools.py#L463">463</a>, <a href="https://github.com/microsoft/agent-framework/blob/c33ec18ae94b050922f73d66a130583346e57dc5/packages/core/agent_framework/_tools.py#L477">477</a>, <a href="https://github.com/microsoft/agent-framework/blob/c33ec18ae94b050922f73d66a130583346e57dc5/packages/core/agent_framework/_tools.py#L484">484</a>, <a href="https://github.com/microsoft/agent-framework/blob/c33ec18ae94b050922f73d66a130583346e57dc5/packages/core/agent_framework/_tools.py#L491">491</a>, <a href="https://github.com/microsoft/agent-framework/blob/c33ec18ae94b050922f73d66a130583346e57dc5/packages/core/agent_framework/_tools.py#L514">514</a>, <a href="https://github.com/microsoft/agent-framework/blob/c33ec18ae94b050922f73d66a130583346e57dc5/packages/core/agent_framework/_tools.py#L516">516</a>, <a href="https://github.com/microsoft/agent-framework/blob/c33ec18ae94b050922f73d66a130583346e57dc5/packages/core/agent_framework/_tools.py#L523">523</a>, <a href="https://github.com/microsoft/agent-framework/blob/c33ec18ae94b050922f73d66a130583346e57dc5/packages/core/agent_framework/_tools.py#L531">531</a>, <a href="https://github.com/microsoft/agent-framework/blob/c33ec18ae94b050922f73d66a130583346e57dc5/packages/core/agent_framework/_tools.py#L650">650</a>, <a href="https://github.com/microsoft/agent-framework/blob/c33ec18ae94b050922f73d66a130583346e57dc5/packages/core/agent_framework/_tools.py#L690-L692">690&ndash;692</a>, <a href="https://github.com/microsoft/agent-framework/blob/c33ec18ae94b050922f73d66a130583346e57dc5/packages/core/agent_framework/_tools.py#L694">694</a>, <a href="https://github.com/microsoft/agent-framework/blob/c33ec18ae94b050922f73d66a130583346e57dc5/packages/core/agent_framework/_tools.py#L700">700</a>, <a href="https://github.com/microsoft/agent-framework/blob/c33ec18ae94b050922f73d66a130583346e57dc5/packages/core/agent_framework/_tools.py#L752-L754">752&ndash;754</a>, <a href="https://github.com/microsoft/agent-framework/blob/c33ec18ae94b050922f73d66a130583346e57dc5/packages/core/agent_framework/_tools.py#L779">779</a>, <a href="https://github.com/microsoft/agent-framework/blob/c33ec18ae94b050922f73d66a130583346e57dc5/packages/core/agent_framework/_tools.py#L805">805</a>, <a href="https://github.com/microsoft/agent-framework/blob/c33ec18ae94b050922f73d66a130583346e57dc5/packages/core/agent_framework/_tools.py#L809">809</a>, <a href="https://github.com/microsoft/agent-framework/blob/c33ec18ae94b050922f73d66a130583346e57dc5/packages/core/agent_framework/_tools.py#L847-L849">847&ndash;849</a>, <a href="https://github.com/microsoft/agent-framework/blob/c33ec18ae94b050922f73d66a130583346e57dc5/packages/core/agent_framework/_tools.py#L853">853</a>, <a href="https://github.com/microsoft/agent-framework/blob/c33ec18ae94b050922f73d66a130583346e57dc5/packages/core/agent_framework/_tools.py#L875">875</a>, <a href="https://github.com/microsoft/agent-framework/blob/c33ec18ae94b050922f73d66a130583346e57dc5/packages/core/agent_framework/_tools.py#L1017-L1023">1017&ndash;1023</a>, <a href="https://github.com/microsoft/agent-framework/blob/c33ec18ae94b050922f73d66a130583346e57dc5/packages/core/agent_framework/_tools.py#L1059">1059</a>, <a href="https://github.com/microsoft/agent-framework/blob/c33ec18ae94b050922f73d66a130583346e57dc5/packages/core/agent_framework/_tools.py#L1071">1071</a>, <a href="https://github.com/microsoft/agent-framework/blob/c33ec18ae94b050922f73d66a130583346e57dc5/packages/core/agent_framework/_tools.py#L1073">1073</a>, <a href="https://github.com/microsoft/agent-framework/blob/c33ec18ae94b050922f73d66a130583346e57dc5/packages/core/agent_framework/_tools.py#L1075">1075</a>, <a href="https://github.com/microsoft/agent-framework/blob/c33ec18ae94b050922f73d66a130583346e57dc5/packages/core/agent_framework/_tools.py#L1078-L1081">1078&ndash;1081</a>, <a href="https://github.com/microsoft/agent-framework/blob/c33ec18ae94b050922f73d66a130583346e57dc5/packages/core/agent_framework/_tools.py#L1102">1102</a>, <a href="https://github.com/microsoft/agent-framework/blob/c33ec18ae94b050922f73d66a130583346e57dc5/packages/core/agent_framework/_tools.py#L1106">1106</a>, <a href="https://github.com/microsoft/agent-framework/blob/c33ec18ae94b050922f73d66a130583346e57dc5/packages/core/agent_framework/_tools.py#L1110">1110</a>, <a href="https://github.com/microsoft/agent-framework/blob/c33ec18ae94b050922f73d66a130583346e57dc5/packages/core/agent_framework/_tools.py#L1124-L1126">1124&ndash;1126</a>, <a href="https://github.com/microsoft/agent-framework/blob/c33ec18ae94b050922f73d66a130583346e57dc5/packages/core/agent_framework/_tools.py#L1467">1467</a>, <a href="https://github.com/microsoft/agent-framework/blob/c33ec18ae94b050922f73d66a130583346e57dc5/packages/core/agent_framework/_tools.py#L1576-L1582">1576&ndash;1582</a>, <a href="https://github.com/microsoft/agent-framework/blob/c33ec18ae94b050922f73d66a130583346e57dc5/packages/core/agent_framework/_tools.py#L1711">1711</a>, <a href="https://github.com/microsoft/agent-framework/blob/c33ec18ae94b050922f73d66a130583346e57dc5/packages/core/agent_framework/_tools.py#L1715">1715</a>, <a href="https://github.com/microsoft/agent-framework/blob/c33ec18ae94b050922f73d66a130583346e57dc5/packages/core/agent_framework/_tools.py#L1761">1761</a>, <a href="https://github.com/microsoft/agent-framework/blob/c33ec18ae94b050922f73d66a130583346e57dc5/packages/core/agent_framework/_tools.py#L1822-L1823">1822&ndash;1823</a>, <a href="https://github.com/microsoft/agent-framework/blob/c33ec18ae94b050922f73d66a130583346e57dc5/packages/core/agent_framework/_tools.py#L1938">1938</a>, <a href="https://github.com/microsoft/agent-framework/blob/c33ec18ae94b050922f73d66a130583346e57dc5/packages/core/agent_framework/_tools.py#L1958">1958</a>, <a href="https://github.com/microsoft/agent-framework/blob/c33ec18ae94b050922f73d66a130583346e57dc5/packages/core/agent_framework/_tools.py#L1960">1960</a>, <a href="https://github.com/microsoft/agent-framework/blob/c33ec18ae94b050922f73d66a130583346e57dc5/packages/core/agent_framework/_tools.py#L2016">2016</a>, <a href="https://github.com/microsoft/agent-framework/blob/c33ec18ae94b050922f73d66a130583346e57dc5/packages/core/agent_framework/_tools.py#L2079">2079</a>, <a href="https://github.com/microsoft/agent-framework/blob/c33ec18ae94b050922f73d66a130583346e57dc5/packages/core/agent_framework/_tools.py#L2251-L2252">2251&ndash;2252</a>, <a href="https://github.com/microsoft/agent-framework/blob/c33ec18ae94b050922f73d66a130583346e57dc5/packages/core/agent_framework/_tools.py#L2272">2272</a>, <a href="https://github.com/microsoft/agent-framework/blob/c33ec18ae94b050922f73d66a130583346e57dc5/packages/core/agent_framework/_tools.py#L2328-L2329">2328&ndash;2329</a>, <a href="https://github.com/microsoft/agent-framework/blob/c33ec18ae94b050922f73d66a130583346e57dc5/packages/core/agent_framework/_tools.py#L2475-L2476">2475&ndash;2476</a>, <a href="https://github.com/microsoft/agent-framework/blob/c33ec18ae94b050922f73d66a130583346e57dc5/packages/core/agent_framework/_tools.py#L2551">2551</a>, <a href="https://github.com/microsoft/agent-framework/blob/c33ec18ae94b050922f73d66a130583346e57dc5/packages/core/agent_framework/_tools.py#L2556">2556</a>, <a href="https://github.com/microsoft/agent-framework/blob/c33ec18ae94b050922f73d66a130583346e57dc5/packages/core/agent_framework/_tools.py#L2563">2563</a></td></tr><tr><td><b>TOTAL</b></td><td><b>29076</b></td><td><b>3467</b></td><td><b>88%</b></td><td>&nbsp;</td></tr></tbody></table></details>

## Python Unit Test Overview
| Tests | Skipped | Failures | Errors | Time |
| ----- | ------- | -------- | -------- | ------------------ |
| 5805 | 30 :zzz: | 0 :x: | 0 :fire: | 1m 37s :stopwatch: |

**Review by chetantoshniwal — COMMENTED** [review]
## Automated Code Review

**Reviewers:** 4 | **Confidence:** 98% | **Result:** All clear

Reviewed: Correctness, Security Reliability, Test Coverage, Design Approach

---
*Automated review by chetantoshniwal's agents*

File: python/packages/core/tests/core/test_function_invocation_logic.py (hunk 1 above)
Comment (chetantoshniwal) [reply python/packages/core/tests/core/test_function_invocation_logic.py:4027]: Fixed in `c33ec18a`.

File: python/packages/core/agent_framework/_tools.py (hunk 2 above)
Comment (chetantoshniwal) [reply python/packages/core/agent_framework/_tools.py]: Fixed in `c33ec18a`.

File: python/packages/core/agent_framework/_tools.py (hunk 3 above)
Comment (chetantoshniwal) [reply python/packages/core/agent_framework/_tools.py]: Fixed in `c33ec18a`.

File: python/packages/core/tests/core/test_function_invocation_logic.py [hunk 4]
```diff
@@ -3807,3 +3807,97 @@ def empty_request(task: str) -> str:
...
+            "tool_choice": "auto",
+            "continuation_token": {"response_id": "resp_abc123"},
+        },
+    )
+
+    assert chat_client_base.call_count >= 2
+    assert "continuation_token" in recorded_options[0], "First call should include continuation_token"
+    assert "continuation_token" not in recorded_options[1], (
+        "continuation_token must be stripped after the background response completes"
+    )
```
[the remark this answers is not in the document]
Comment (chetantoshniwal) [reply python/packages/core/tests/core/test_function_invocation_logic.py]: Fixed in `c33ec18a`.

File: python/packages/core/agent_framework/_tools.py [hunk 5]
```diff
@@ -2349,6 +2349,14 @@ async def _get_response() -> ChatResponse[Any]:
                         options=mutable_options,
                     )
 
+                    # Once a background response completes, strip polling/background
+                    # options so subsequent tool-loop iterations POST results normally.
+                    if response.continuation_token is None and (
+                        "continuation_token" in mutable_options or "background" in mutable_options
+                    ):
+                        mutable_options.pop("continuation_token", None)
+                        mutable_options.pop("background", None)
```
Comment (eavanvalkenburg) [inline python/packages/core/agent_framework/_tools.py:2358]: I don't know if this is right, the intent of running in the background is that the whole run does, now it runs background until a tool call and then goes non-background after the tool execution, that seems unintented.

**Comment by eavanvalkenburg** [comment]
This was fixed in another PR, closing.
